js-packages/@prestojs/rest/src/useAsyncValue.ts

```diff
@@ -0,0 +1,232 @@
+import { getId as _identifiableGetId, hashId, Id, Identifiable, useAsync } from '@prestojs/util';
+import isEqual from 'lodash/isEqual';
+import { useEffect, useRef } from 'react';
+
+/**
+ * Wrapper around default implementation so we can give more specific error message
+ */
+function identifiableGetId(item: Identifiable | any, fallbackGetId?: (item: any) => Id): Id {
+    try {
+        return _identifiableGetId(item, fallbackGetId);
+    } catch (err) {
+        throw new Error(err.message + ' To fix pass `getId` to `useAsyncValue`.');
+    }
+}
+
+type CommonProps<T, U extends Id> = {
+    /**
+     * An optional array of existing values to try and find the value in.
+     *
+     * If each item does not implement [Identifiable](doc:Identifiable) then you
+     * must provide the `getId` function.
+     *
+     * You can use this to avoid resolving data that already exists. If dealing with
+     * ViewModel instances you can use it with [useViewModelCache](doc:useViewModelCache),
+     * eg.
+     *
+     * ```js
+     * const existingValues = useViewModelCache(User, cache => cache.get(id, fieldNames));
+     * const { value } = useAsyncValue({ id, existingValues, resolve: fetchUser });
+     * ```
+     */
+    existingValues?: T[];
```

Comment:
I never asked this, but why is this an array rather than a mapping/dict?

If you had a large number of records this would become quite slow, or is the point that you shouldn't be passing in that many records anyway because it's unlikely you really have that many dependencies and it will cause unnecessary rerenders?

Reply:
Main reason is for ease of use with `useAsyncLookup` (you can pass the result of that straight through as `existingValues`). You'll also mostly have an array so passing a mapping/dict requires you to transform it to an object anyway (ie. you'd have to iterate over it all once anyway).

The other reason is compound keys wouldn't work as keys in either Map or object (would have to be stringified first)